RavenDB-5934 Implement encryption

iftahbe opened the pull request.

ayende commented on src/Voron/Impl/Journal/JournalReader.cs:
```diff
@@ -121,16 +126,22 @@ public bool ReadOneTransactionToDataFile(StorageEnvironmentOptions options)
+                // There is a chance the page was not synced to the data file before we crashed.
+                // If encryption is enabled, when recovering, we cannot decrypt an empty (zeroed) page
+                var pagePtr = options.EncryptionEnabled 
+                    ? _dataPager.AcquirePagePointerForNewPage(this, pageInfoPtr[i].PageNumber, numberOfPagesOnDestination) 
```
Probably want to always have a new page here, and need to explain that we also do that for verification purposes, because we are going to overwrite the page anyway, so we don't care about its current contents

ayende commented on src/Voron/Impl/Journal/JournalReader.cs:
```diff
@@ -121,16 +126,22 @@ public bool ReadOneTransactionToDataFile(StorageEnvironmentOptions options)
 
-                if (pageInfoPtr[i].DiffSize == 0)
+                // When using encryption we copy the entire page from the journal, not just the diff. 
+                if (pageInfoPtr[i].DiffSize == 0 || options.EncryptionEnabled)
```
When using encryption, we already have `DiffSize == 0`, condition redundant.

ayende commented on src/Voron/Impl/Journal/JournalReader.cs:
```diff
@@ -160,11 +172,53 @@ public void RecoverAndValidate(StorageEnvironmentOptions options)
+                return;
+            var compressionBufferSize = _recoveryPager.NumberOfAllocatedPages * Constants.Storage.PageSize;
+            var pagePointer = _recoveryPager.AcquirePagePointer(tx, 0);
+            Memory.Set(pagePointer, 0, compressionBufferSize);
```
Use Sodium Zero

ayende commented on src/Voron/Impl/Journal/JournalReader.cs:
```diff
@@ -160,11 +172,53 @@ public void RecoverAndValidate(StorageEnvironmentOptions options)
+            var num = txHeader->TransactionId;
+            ulong macLen = 16;
+
+            if (txHeader->Flags != EncryptionFlags.Encrypted)
```
This isn't flags, need to treat it as flags.

ayende commented on src/Voron/Impl/Journal/TransactionHeader.cs:
```diff
@@ -43,7 +46,7 @@ public struct TransactionHeader
 
         [FieldOffset(36)]
-        public int Reserved1;
+        public EncryptionFlags Flags;
```
Use TransactionFlags, not Encryption Flags

ayende commented on src/Voron/EncryptionFlags.cs:
```diff
@@ -0,0 +1,7 @@
+namespace Voron
+{
+    public enum EncryptionFlags
```
Make sure to use `[Flags]`

ayende commented on src/Voron/Impl/Journal/JournalReader.cs:
```diff
@@ -204,6 +258,36 @@ private bool TryReadAndValidateHeader(StorageEnvironmentOptions options, out Tra
+                //  to hold the transaction before decrypting, and release the buffers afterwards.
+                var size = (4*Constants.Size.Kilobyte) * GetNumberOf4KbFor(sizeof(TransactionHeader) + current->CompressedSize);
+
+                var buffer = new EncryptionBuffer
```
Probably need to pool those and register them in the `NativeMemory` tracking

ayende commented on src/Voron/Impl/Journal/JournalReader.cs:
```diff
@@ -293,25 +377,40 @@ public override string ToString()
+            if (_encryptionBuffers != null) // Encryption enabled
+			{ 
+                foreach (var buffer in _encryptionBuffers)
+                    Win32MemoryProtectMethods.VirtualFree(buffer.Pointer, UIntPtr.Zero, Win32MemoryProtectMethods.FreeType.MEM_RELEASE);
```
You just broke our Linux support, because this will try to call Win32 methods directly

ayende commented on src/Voron/Impl/Journal/TransactionHeader.cs:
```diff
@@ -66,6 +69,17 @@ public struct TransactionHeader
         [FieldOffset(128)]
         public long TimeStampTicksUtc; // DateTime.UtcNow.Ticks when the tx happened
 
+        [FieldOffset(136)]
```
Did you change the schema version here?

ayende commented on src/Voron/Impl/Journal/WriteAheadJournal.cs:
```diff
@@ -1274,38 +1276,61 @@ private CompressedPagesResult PrepareToWriteToJournal(LowLevelTransaction tx)
+
+                    write += size;
+                    pagesInfo[pageSequencialNumber].Size = size;
+                    pagesInfo[pageSequencialNumber].DiffSize = -1;
```
Need to use 0 here

ayende commented on src/Voron/Impl/Journal/WriteAheadJournal.cs:
```diff
@@ -1460,6 +1527,17 @@ public void ReduceSizeOfCompressionBufferIfNeeded()
+
+            var compressionBufferSize = _compressionPager.NumberOfAllocatedPages * Constants.Storage.PageSize;
+            var pagePointer = _compressionPager.AcquirePagePointer(tx, 0);
+            Memory.Set(pagePointer, 0, compressionBufferSize);
```
Sodium zero

ayende commented on src/Voron/Impl/Paging/CryptoPager.cs:
```diff
@@ -1,21 +1,109 @@
+        public CryptoPager(AbstractPager inner) : base(inner.Options, inner.UsePageProtection)
         {
-            _masterKey = Sodium.GenerateMasterKey();
+
```
Copy `FileName = inner.FileName`

ayende replied:
Options are the same and shouldn't be redirected

ayende commented on src/Voron/Impl/Paging/CryptoPager.cs:
```diff
@@ -27,75 +115,308 @@ private static int GetNumberOfPages(PageHeader* header)
+
+        public override byte* AcquirePagePointer(IPagerLevelTransactionState tx, long pageNumber, PagerState pagerState = null)
+        {
+            _readerWriterLock.EnterReadLock();
```
Should be able to remove this

ayende commented on src/Voron/Impl/Paging/CryptoPager.cs:
```diff
@@ -27,75 +115,308 @@ private static int GetNumberOfPages(PageHeader* header)
+            encBuffer.Size = Constants.Storage.PageSize;
+        }
+
+        public EncryptionBuffer GetNewBufferAndAddToTx(long pageNumber, CryptoTransactionState state, int size)
```
Can be made private

ayende commented on src/Voron/Impl/Paging/CryptoPager.cs:
```diff
@@ -27,75 +115,308 @@ private static int GetNumberOfPages(PageHeader* header)
+            }
+            finally
+            {
+                _readerWriterLock.ExitWriteLock();
```
can remove this

ayende commented on src/Voron/Impl/Paging/CryptoPager.cs:
```diff
@@ -27,75 +115,308 @@ private static int GetNumberOfPages(PageHeader* header)
 
-            ByteString subKey;
-            using (allocator.Allocate(32, out subKey))
+            if (tx.IsWriteTransaction)
```
why only for write transactions?

ayende commented on src/Voron/Impl/Paging/CryptoPager.cs:
```diff
@@ -27,75 +115,308 @@ private static int GetNumberOfPages(PageHeader* header)
+            if (Disposed)
+                return;
+
+            Disposed = true;
```
Missing `GC.SuppressFinalizer` and do we have any unmanaged resource we manage that we can't let the inner one handle it?

ayende commented on src/Voron/Impl/Paging/CryptoPager.cs:
```diff
@@ -27,75 +115,308 @@ private static int GetNumberOfPages(PageHeader* header)
+            return new Crypto4KbBatchWrites(this);
+        }
+
+        private class Crypto4KbBatchWrites : I4KbBatchWrites
```
delete this class

ayende commented on src/Voron/Impl/Paging/CryptoPager.cs:
```diff
@@ -27,75 +115,308 @@ private static int GetNumberOfPages(PageHeader* header)
+
+        public override I4KbBatchWrites BatchWriter()
+        {
+            return new Crypto4KbBatchWrites(this);
```
Forward to the inner, instead

ayende commented on src/Voron/Impl/Scratch/ScratchBufferFile.cs:
```diff
@@ -275,11 +280,16 @@ public Page ReadPage(LowLevelTransaction tx, long p, PagerState pagerState = nul
+            return _scratchPager.AcquirePagePointerForNewPage(tx, p, numberOfPages);
         }
 
+        public void BreakLargeAllocationToSeparatePages(IPagerLevelTransactionState tx, long p)
```
Forward this to the pager always, then it can decide what to do

ayende commented on src/Voron/PageHeader.cs:
```diff
@@ -12,6 +12,8 @@ namespace Voron.Data
     public unsafe struct PageHeader
     {
         public static int ChecksumOffset = (int)Marshal.OffsetOf<PageHeader>(nameof(Checksum));
+        public static int NonceOffset = (int)Marshal.OffsetOf<PageHeader>(nameof(Nonce));
```
Comment on why they are those specific offsets

ayende commented on src/Voron/StorageEnvironmentOptions.cs:
```diff
@@ -521,47 +529,35 @@ public void DeleteAllTempBuffers()
             }
 
-            public override AbstractPager CreateScratchPager(string name, long initialSize)
+            public override AbstractPager CreateTemporaryBufferPager(string name, long initialSize)
```
Comment on why this is special

ayende commented on src/Voron/StorageEnvironmentOptions.cs:
```diff
@@ -521,47 +529,35 @@ public void DeleteAllTempBuffers()
                 bool deleteOnClose = false,
                 bool usePageProtection = false)
             {
+                return EncryptionEnabled
```
Use `if` here, clearer

ayende commented on src/Voron/StorageEnvironmentOptions.cs:
```diff
@@ -764,6 +760,15 @@ public override unsafe void WriteHeader(string filename, FileHeader* header)
 
             private AbstractPager GetTempMemoryMapPager(PureMemoryStorageEnvironmentOptions options, string path, long? intialSize, Win32NativeFileAttributes win32NativeFileAttributes)
             {
+                return EncryptionEnabled
```
use `if`

ayende commented on src/Voron/StorageEnvironmentOptions.cs:
```diff
@@ -779,17 +784,20 @@ private AbstractPager GetTempMemoryMapPager(PureMemoryStorageEnvironmentOptions
             }
 
-            public override AbstractPager CreateScratchPager(string name, long intialSize)
+            public override AbstractPager CreateTemporaryBufferPager(string name, long initialSize)
```
comment why needed

ayende commented on src/Voron/StorageEnvironmentOptions.cs:
```diff
@@ -779,17 +784,20 @@ private AbstractPager GetTempMemoryMapPager(PureMemoryStorageEnvironmentOptions
+                return GetTempMemoryMapPagerInternal(this, Path.Combine(TempPath, filename), initialSize, Win32NativeFileAttributes.RandomAccess | Win32NativeFileAttributes.DeleteOnClose | Win32NativeFileAttributes.Temporary);
             }
 
             public override AbstractPager OpenPager(string filename)
```
Need to make sure that we just wrap this method call, instead

ayende commented on src/Voron/Util/DataCopier.cs:
```diff
@@ -38,6 +38,11 @@ public void ToStream(byte* ptr, long count, Stream output)
         {
+            // In case of encryption bundle, we don't want to decrypt the data for backup, 
+            // so let's work directly with the underlying encrypted data (Inner pager).
+            if (src is CryptoPager)
```
Bad, leak impl details. Make use of something like the `BatchWriter`, which can already handle this.

iftahbe committed:
RavenDB-5934 Implement encryption
- Encryption is turned off until we do key management
- Added CryptoPager
- Added CryptoPagerTransactionState to manage buffers in the transaction
- Added encryption fields "Nonce" and "Mac" to PageHeader and TransactionHeader
- Turned off async commit, journal diffs and checksum validation when using encryption

iftahbe committed:
RavenDB-5934-encryption libsodium changes

iftahbe committed:
RavenDB-5934 No actions are allowed between tx commit and dispose

iftahbe committed:
RavenDB-5934 Fixed totalSizeWritten little bug in PrepareToWriteToJournal()

iftahbe committed:
RavenDB-5934 turning encryption off

iftahbe committed:
RavenDB-5934 fix PR comments

redknightlois commented on src/Sparrow/Platform/UnmanagedMemory.cs:
```diff
@@ -43,5 +46,36 @@ public static IntPtr Set(byte* dest, int c, long count)
         }
+
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        public static byte* Allocate4KAllignedMemory(long size)
```
Typo here ```Allocate4KAlignedMemory```

redknightlois commented:
Quick question: Will the unprotected transaction data be written to a disk backed pager at any point in time? Couldnt figure that out from the code itself, but it is a huge point.

ayende commented on src/Sparrow/Platform/Posix/Syscall.cs:
```diff
@@ -78,7 +78,14 @@ public static unsafe long pread(int fd, void* buf, ulong count, long offset)
+		// posix_memalign(3)
+        //     int posix_memalign(void** memptr, size_t alignment, size_t size);
+        [DllImport(LIBC_6, SetLastError = true)]
+        public static extern int posix_memalign(IntPtr pPtr, long allignment, long count);
```
`size_t` is actually `IntPtr`, need to change it (would fail on the Pi)

ayende commented on src/Sparrow/Platform/UnmanagedMemory.cs:
```diff
@@ -43,5 +46,36 @@ public static IntPtr Set(byte* dest, int c, long count)
+            {
+                var p = new IntPtr();
+                var pp = new IntPtr(&p);
+                Syscall.posix_memalign(pp, 4096, size);
```
This is a very awkward way of doing this.

ayende replied:
Where is the error handling?

ayende commented:
No, the data is always encrypted on the pagers, we copy the encrypted data to our own memory (currently just `VirtualAlloc`) and decrypt there.
At no point should you have the plain data on the pagers.

The buffers we use are secruely zeroed after use, so the decrypted data is only there for the lifetime of the transaction, once the tx is complete, the data is gone.

That doesn't mean that you can't peek into the decrypted data if you can access the memory, or force the data to go the page file, etc.

We do have a few pagers, (compression, recovery, etc) that are used for unencrypted data, and this [issue](http://issues.hibernatingrhinos.com/issue/RavenDB-6799) will track making sure that we are properly zeroing all the data.
